### PR TITLE
Enhance training dataset format support

### PR DESCRIPTION
**Pull request by talmo**
# Overview

**Goal:** Be able to store SLEAP package files ([example](https://colab.research.google.com/drive/1QlC-zk5AqklGO-PEzo7dwShFXYGhKLul#scrollTo=id5JzQL9_KIS)) in `PoseTraining` objects so we can embed images used for training models together with the ground truth poses.

## Data model equivalences

| `sleap-io` | `ndx-pose` |
|--------|--------|
| `Labels` | `PoseTraining` |
| `LabeledFrame` | `TrainingFrame` |
| `Skeleton` | `Skeleton` |
| `Instance` | `SkeletonInstance` |
| `Video` | `ImageSeries` |
| `Labels.skeletons` | `Skeletons` |
| `Labels.labeled_frames` | `TrainingFrames` |
| `Labels.videos` | `SourceVideos` |

### MVP

- [x] Make `TrainingFrame.skeleton_instance` into `TrainingFrame.skeleton_instances` so it can hold multiple instances per frame (https://github.com/talmolab/ndx-pose/pull/3)
- [x] Put `SkeletonInstance` creation inside of `TrainingFrame` loop in the README example
- [x] Store image data in [`Image`](https://nwb-schema.readthedocs.io/en/latest/format.html#image) as 3-D array (https://github.com/talmolab/ndx-pose/pull/2)
- [x] Make `TrainingFrame.source_video` optional so we don't need to keep the source video together with the NWB file (https://github.com/talmolab/ndx-pose/pull/2)
- [ ] Integrate into [`sleap_io.io.nwb`](https://github.com/talmolab/sleap-io/blob/main/sleap_io/io/nwb.py)
- [ ] Update docs/readme
- [x] Finish tests

### Nice to have

- [ ] Add a plain string field to `TrainingFrame` to point to source video as provenance but without triggering validation on DANDI
- [ ] Add support for PNG string encoding representation for `Image`
- [ ] Convert and upload [reference SLEAP training datasets](https://sleap.ai/datasets.html) to DANDI
- [ ] Add support for directly opening SLEAP Dandisets in `sleap-io` via [`dandi`](https://dandi.readthedocs.io/en/latest/modref/generated/dandi.download.html#module-dandi.download) API

### CODE REVIEW

File: spec/ndx-pose.extensions.yaml [hunk 1]
```diff
@@ -159,8 +159,8 @@ groups:
 - neurodata_type_def: SkeletonInstance
   neurodata_type_inc: NWBDataInterface
   default_name: skeleton_instance
-  doc: Group that holds ground-truth pose data for a single instance of a skeleton
-    in a single frame. This is meant to be used within a TrainingFrame.
+  doc: 'Group that holds ground-truth pose data for a single instance of a skeleton
+    in a single frame. '
```
Comment (rly) [inline spec/ndx-pose.extensions.yaml:163]:
```suggestion
    in a single frame.'
```

File: spec/ndx-pose.extensions.yaml [hunk 2]
```diff
@@ -200,6 +200,16 @@ groups:
   - neurodata_type_inc: TrainingFrame
     doc: Ground-truth position data for all instances of a skeleton in a single frame.
     quantity: '*'
+- neurodata_type_def: SkeletonInstances
+  neurodata_type_inc: NWBDataInterface
+  default_name: skeleton_instances
+  doc: Organizational group to hold skeleton instances.This is meant to be used within
```
Comment (rly) [inline spec/ndx-pose.extensions.yaml:206]:
```suggestion
  doc: Organizational group to hold skeleton instances. This is meant to be used within
```

File: src/pynwb/ndx_pose/testing/mock/pose.py [hunk 3]
```diff
@@ -166,20 +198,35 @@ def mock_source_video(
...
+    *,
+    name: Optional[str] = None,
+):
+    return RGBImage(name=name, data=np.random.rand(640, 480, 3).astype("uint8"))
+
+def mock_source_frame(
+    *,
+    name: Optional[str] = None,
+):
+    return RGBImage(name=name, data=np.random.rand(640, 480, 3).astype("uint8"))
```
Comment (rly) [inline src/pynwb/ndx_pose/testing/mock/pose.py:211]:
```suggestion
```

Comment (rly) [reply src/pynwb/ndx_pose/testing/mock/pose.py:211]: Already defined above

**Comment by rly** [comment]
Hi @talmo and @eberrigan, thank you for all your work on this PR and for looking through the recent changes in the main branch at brainhack! I am happy to take on updating the README example and test, but I think you all would be better suited to do `Integrate into sleap_io.io.nwb` and `Update docs/readme`. Let me know if you'd like me to jump in to update the README example and test.

**Comment by eberrigan** [comment]
Feel free to update the readme. Thanks so much!

**Comment by rly** [comment]
@talmo Can you give me write access to your fork? It will be easier than me creating a fork of your fork and us having to merge each fork in one by one...

**Comment by talmo** [comment]
> @talmo Can you give me write access to your fork? It will be easier than me creating a fork of your fork and us having to merge each fork in one by one...

Done! https://github.com/talmolab/ndx-pose/invitations

**Comment by rly** [comment]
Thanks! I made a few changes - I moved the `Skeletons` type out of the `PoseTraining` container type because the `PoseEstimation` type now requires a `Skeleton` object and, for people who don't have (or want to include) training data with their pose estimates, it feels weird in my opinion, to require them to make a `PoseTraining` container just to put `Skeletons` into it.

I also updated the readme and tests and made some other minor adjustments.

@talmo I would like to merge this PR soon so that it will be easier for me to refactor the repo to use the latest ndx-template. Then you could start a new PR with any necessary adjustments for integration with SLEAP and any other changes you want to make (or you can make those changes here if you have time). Would that be all right?

**Comment by talmo** [comment]
Yes that works for me @rly!
